handlers/handlers.go

```diff
@@ -57,6 +65,59 @@ func wrapGetWithErrors(ef func(error), hf handlerFunc) http.Handler {
 	})
 }
 
+func hmacAuth(hf handlerFunc, secretKey string, serviceId string) handlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) (int, error) {
+		// If secret key isn't set, allow all requests
+		if secretKey == "" {
+			return hf(w, r)
+		}
+
+		var rawSignature, rawSignDate string
+
+		query := r.URL.Query()
+
+		if signature := query.Get("signature"); signature != "" {
+			rawSignature = signature
+		} else if signature := r.Header.Get("X-Signature"); signature != "" {
+			rawSignature = signature
+		} else {
+			return 400, errors.New("No signature provided")
+		}
+
+		if date := query.Get("date"); date != "" {
+			rawSignDate = date
+		} else if date := r.Header.Get("X-Signature-Date"); date != "" {
+			rawSignDate = date
+		} else {
+			return 400, errors.New("No signature date provided")
+		}
+
+		signDate, err := time.Parse(time.RFC3339Nano, rawSignDate)
+		if err != nil || time.Now().Sub(signDate) > maxSignatureAge {
```

Comment:
I think it would be a good idea to split out the error case from the expired case.  Expired should probably return a `401` error instead of `400`, because the request parameters themselves are valid, just no longer authorized.

Reply:
Yes, good point. I meant to do that, actually.